README
# checking pull-requests feature

"git" can mean anything, depending on your mood.

 - random three-letter combination that is pronounceable, and not
   actually used by any common UNIX command.  The fact that it is a
   mispronunciation of "get" may or may not be relevant.
 - stupid. contemptible and despicable. simple. Take your pick from the
   dictionary of slang.
 - "global information tracker": you're in a good mood, and it actually
   works for you. Angels sing, and a light suddenly fills the room.
 - "goddamn idiotic truckload of sh*t": when it breaks

Git is a fast, scalable, distributed revision control system with an
unusually rich command set that provides both high-level operations
and full access to internals.

Git is an Open Source project covered by the GNU General Public License.
It was originally written by Linus Torvalds with help of a group of
hackers around the net. It is currently maintained by Junio C Hamano.

Please read the file INSTALL for installation instructions.

See Documentation/gittutorial.txt to get started, then see
Documentation/everyday.txt for a useful minimum set of commands, and
Documentation/git-commandname.txt for documentation of each command.
If git has been correctly installed, then the tutorial can also be
read with "man gittutorial" or "git help tutorial", and the
documentation of each command with "man git-commandname" or "git help
commandname".

CVS users may also want to read Documentation/gitcvs-migration.txt
("man gitcvs-migration" or "git help cvs-migration" if git is
installed).

Many Git online resources are accessible from http://git-scm.com/
including full documentation and Git related tools.

The user discussion and development of Git take place on the Git
mailing list -- everyone is welcome to post bug reports, feature
requests, comments and patches to git@vger.kernel.org. To subscribe
to the list, send an email with just "subscribe git" in the body to
majordomo@vger.kernel.org. The mailing list archives are available at
http://marc.theaimsgroup.com/?l=git and other archival sites.

The messages titled "A note from the maintainer", "What's in
git.git (stable)" and "What's cooking in git.git (topics)" and
the discussion following them on the mailing list give a good
reference for project status, development direction and
remaining tasks.


debian/changelog
git (1:1.9.1-1.1) unstable; urgency=low

  * package git-credential-gnome-keyring (closes: #740739)
  * debian/diff/0010-contrib-credential-gnome-keyring-Makefile....diff:
    new; add install target
  * debian/rules: add PKG_ARCH list to build arch-dependent packages

 -- Philipp A. Hartmann <pah@qo.cx>  Fri, 21 Mar 2014 17:31:26 +0100

git (1:1.9.1-1) unstable; urgency=low

  * new upstream point release.

  [ Jonathan Nieder ]
  * git-email: Recommends: libio-socket-ssl-perl (>= 1.951) since
    earlier versions do not use OpenSSL's defaults when ca_path
    and ca_file are unset.

  [ Gilles Filippini ]
  * install git-subtree from contrib (closes: #704652).
  * debian/diff/0009-contrib-subtree-unset-prefix-before-proceeding.diff:
    new; unset the 'prefix' environment variable before proceeding.

 -- Jonathan Nieder <jrnieder@gmail.com>  Tue, 18 Mar 2014 15:58:46 -0700

git (1:1.9.0-1) unstable; urgency=low

  * new upstream release.

 -- Jonathan Nieder <jrnieder@gmail.com>  Tue, 18 Feb 2014 08:31:49 -0800

git (1:1.9.0~rc3-1) unstable; urgency=low

  * new upstream release candidate.

 -- Jonathan Nieder <jrnieder@gmail.com>  Fri, 07 Feb 2014 15:53:52 -0800

git (1:1.9~rc1-1) unstable; urgency=low

  * new upstream release candidate.

 -- Jonathan Nieder <jrnieder@gmail.com>  Mon, 27 Jan 2014 14:07:56 -0800

git (1:1.9~rc0-1) unstable; urgency=low

  * new upstream release candidate.
  * debian/diff:
    * 0001-remove-interpreter-line-from-shell-libraries.diff:
      remove; applied upstream.
    * 0002...0009: rename to 0001-*, ..., 0008-*.
    * 0010-git-remote-mediawiki-do-not-remove-installed-...diff:
      remove; applied upstream.
  * debian/rules install: no more need to include DESTDIR in
    mw-to-git INSTLIBDIR.

 -- Jonathan Nieder <jrnieder@gmail.com>  Mon, 27 Jan 2014 13:05:17 -0800

git (1:1.8.5.3-1) unstable; urgency=low

  * new upstream point release.

 -- Jonathan Nieder <jrnieder@gmail.com>  Thu, 16 Jan 2014 16:18:25 -0800

git (1:1.8.5.2-2) unstable; urgency=low

  * /etc/bash_completion.d/git-prompt: check for git-sh-prompt
    and do not source it if git is removed but not purged (thx
    Craig Small; closes: #726669).
  * debian/control: package git Breaks: git-buildpackage versions
    that rely on output from 'git status' without the --porcelain
    flag (closes: #734223).

 -- Jonathan Nieder <jrnieder@gmail.com>  Thu, 09 Jan 2014 14:46:23 -0800

git (1:1.8.5.2-1) unstable; urgency=low

  * new upstream point release.

 -- Jonathan Nieder <jrnieder@gmail.com>  Tue, 17 Dec 2013 16:31:01 -0800

git (1:1.8.5.1-1) unstable; urgency=medium

  * new upstream point release.
    * submodule: do not copy unknown update mode from .gitmodules.

 -- Jonathan Nieder <jrnieder@gmail.com>  Tue, 03 Dec 2013 14:12:55 -0800

git (1:1.8.5-1) unstable; urgency=low

  * new upstream release.
  * merge branch debian-experimental.
  * git-daemon-sysvinit: advertise GIT_DAEMON_BASE_PATH in
    /etc/default/git-daemon (thx Alexander Dahl and Antonio
    Ospite; closes: #704985).

 -- Jonathan Nieder <jrnieder@gmail.com>  Sat, 30 Nov 2013 11:28:02 -0800

git (1:1.8.5~rc3-1) experimental; urgency=low

  * new upstream release candidate.

  [ Anders Kaseorg ]
  * debian/rules: pass $(OPTS) to contrib/mw-to-git/Makefile
    (closes: #729763).

 -- Jonathan Nieder <jrnieder@gmail.com>  Thu, 21 Nov 2013 16:35:29 -0800

git (1:1.8.5~rc2-1) experimental; urgency=low

  * merge branch debian-sid.
  * new upstream release candidate.

  [ Thorsten Glaser ]
  * package mw-to-git (closes: #718395).
    * debian/control: new package git-mediawiki; Depends:
      libmediawiki-api-perl (>= 0.39), liblwp-protocol-https-perl,
      libdatetime-format-iso8601-perl; package git now Suggests
      and git-all Depends: git-mediawiki.
    * debian/rules: build and install Git::Mediawiki perl module,
      mediawiki remote helper, and 'git mw' command in
      git-mediawiki package.
    * debian/diff/0010-git-remote-mediawiki-do-not-remove-...diff:
      new from upstream: mw-to-git: do not remove installed files
      in "clean" target.
    * debian/git-mediawiki.docs: install git-remote-mediawiki.txt.

  [ Anders Kaseorg ]
  * debian/diff/0003-transport-expose-git_tcp_connect-...diff:
    include tcp.h before using tcp functions (closes: #728508).

  [ Jonathan Nieder ]
  * debian/diff/0012-cvsserver-Determinize-output-to-...diff:
    remove; applied upstream.

 -- Jonathan Nieder <jrnieder@gmail.com>  Wed, 13 Nov 2013 16:12:29 -0800


git (1:1.8.4.4-1) unstable; urgency=low

  * new upstream point release.

 -- Jonathan Nieder <jrnieder@gmail.com>  Thu, 21 Nov 2013 17:51:36 -0800

git (1:1.8.4.3-1) unstable; urgency=low

  * new upstream point release.

 -- Jonathan Nieder <jrnieder@gmail.com>  Mon, 11 Nov 2013 15:51:10 -0800

git (1:1.8.5~rc0-1) experimental; urgency=low

  * new upstream release candidate.
  * debian/diff:
    * 0001-hooks-post-receive-email-set-encoding-to-utf-8.diff,
      0002-post-receive-email-defend-against-non-utf8-...diff:
      remove; applied upstream.
    * 0003...0012: rename to 0001-*, ..., 0010-*.

 -- Jonathan Nieder <jrnieder@gmail.com>  Wed, 30 Oct 2013 16:01:15 -0700

git (1:1.8.4.2-1) unstable; urgency=low

  * new upstream point release.
  * debian/diff/0012-cvsserver-Determinize-output-to-...diff:
    new from upstream: sort cvsserver output to combat Perl 5.18
    hash randomization (thx Anders Kaseorg; closes: #727226).

 -- Jonathan Nieder <jrnieder@gmail.com>  Wed, 30 Oct 2013 16:00:07 -0700

git (1:1.8.4.1-1) unstable; urgency=low

  * merge branch debian-experimental.
  * new upstream point release.
  * debian/control: Build-Depends tcl instead of tcl8.5
    (closes: #725961).

 -- Jonathan Nieder <jrnieder@gmail.com>  Mon, 21 Oct 2013 11:46:45 -0700

git (1:1.8.4-1) experimental; urgency=low

  * new upstream release.

 -- Jonathan Nieder <jrnieder@gmail.com>  Sat, 24 Aug 2013 14:50:38 -0700

git (1:1.8.4~rc4-1) experimental; urgency=low


hex.c
#include "cache.h"

const signed char hexval_table[256] = {
	 -1, -1, -1, -1, -1, -1, -1, -1,		/* 00-07 */
	 -1, -1, -1, -1, -1, -1, -1, -1,		/* 08-0f */
	 -1, -1, -1, -1, -1, -1, -1, -1,		/* 10-17 */
	 -1, -1, -1, -1, -1, -1, -1, -1,		/* 18-1f */
	 -1, -1, -1, -1, -1, -1, -1, -1,		/* 20-27 */
	 -1, -1, -1, -1, -1, -1, -1, -1,		/* 28-2f */
	  0,  1,  2,  3,  4,  5,  6,  7,		/* 30-37 */
	  8,  9, -1, -1, -1, -1, -1, -1,		/* 38-3f */
	 -1, 10, 11, 12, 13, 14, 15, -1,		/* 40-47 */
	 -1, -1, -1, -1, -1, -1, -1, -1,		/* 48-4f */
	 -1, -1, -1, -1, -1, -1, -1, -1,		/* 50-57 */
	 -1, -1, -1, -1, -1, -1, -1, -1,		/* 58-5f */
	 -1, 10, 11, 12, 13, 14, 15, -1,		/* 60-67 */
	 -1, -1, -1, -1, -1, -1, -1, -1,		/* 68-67 */
	 -1, -1, -1, -1, -1, -1, -1, -1,		/* 70-77 */
	 -1, -1, -1, -1, -1, -1, -1, -1,		/* 78-7f */
	 -1, -1, -1, -1, -1, -1, -1, -1,		/* 80-87 */
	 -1, -1, -1, -1, -1, -1, -1, -1,		/* 88-8f */
	 -1, -1, -1, -1, -1, -1, -1, -1,		/* 90-97 */
	 -1, -1, -1, -1, -1, -1, -1, -1,		/* 98-9f */
	 -1, -1, -1, -1, -1, -1, -1, -1,		/* a0-a7 */
	 -1, -1, -1, -1, -1, -1, -1, -1,		/* a8-af */
	 -1, -1, -1, -1, -1, -1, -1, -1,		/* b0-b7 */
	 -1, -1, -1, -1, -1, -1, -1, -1,		/* b8-bf */
	 -1, -1, -1, -1, -1, -1, -1, -1,		/* c0-c7 */
	 -1, -1, -1, -1, -1, -1, -1, -1,		/* c8-cf */
	 -1, -1, -1, -1, -1, -1, -1, -1,		/* d0-d7 */
	 -1, -1, -1, -1, -1, -1, -1, -1,		/* d8-df */
	 -1, -1, -1, -1, -1, -1, -1, -1,		/* e0-e7 */
	 -1, -1, -1, -1, -1, -1, -1, -1,		/* e8-ef */
	 -1, -1, -1, -1, -1, -1, -1, -1,		/* f0-f7 */
	 -1, -1, -1, -1, -1, -1, -1, -1,		/* f8-ff */
};

static int get_hex(const char *hex, unsigned char *res, int len)
{
	int i;
	for (i = 0; i < len; i++) {
		unsigned int val;
		/*
		 * hex[1]=='\0' is caught when val is checked below,
		 * but if hex[0] is NUL we have to avoid reading
		 * past the end of the string:
		 */
		if (!hex[0])
			return -1;
		val = (hexval(hex[0]) << 4) | hexval(hex[1]);
		if (val & ~0xff)
			return -1;
		*res++ = val;
		hex += 2;
	}
	return 0;
}

int get_sha1_hex(const char *hex, unsigned char *sha1)
{ return get_hex(hex, sha1, 20); }

int get_md5_hex(const char *hex, unsigned char *md5)
{ return get_hex(hex, md5, 16); }

static char *to_hex(const unsigned char *hash, int len)
{
	static int bufno;
	static char hexbuffer[4][50];
	static const char hex[] = "0123456789abcdef";
	char *buffer = hexbuffer[3 & ++bufno], *buf = buffer;
	int i;

	for (i = 0; i < len; i++) {
		unsigned int val = *hash++;
		*buf++ = hex[val >> 4];
		*buf++ = hex[val & 0xf];
	}
	*buf = '\0';

	return buffer;
}

char *sha1_to_hex(const unsigned char *sha1)
{ return to_hex(sha1, 20); }

char *md5_to_hex(const unsigned char *md5)
{ return to_hex(md5, 16); }
